openmls/src/group/core_group/create_commit.rs

```diff
@@ -107,6 +107,23 @@ impl CoreGroup {
             diff.set_own_index(own_leaf_index);
         }
 
+        // Validate the proposals by doing the following checks:
+
+        // ValSem100
+        // ValSem101
+        // ValSem102
+        // ValSem103
+        // ValSem104
+        // ValSem105
+        // ValSem106
+        self.validate_add_proposals(&proposal_queue)?;
+        // ValSem107
+        // ValSem108
+        self.validate_remove_proposals(&proposal_queue)?;
+        // ValSem109
+        // ValSem110
+        self.validate_update_proposals(&proposal_queue, None)?;
```

Comment:
I'm wondering if we shouldn't do some of these validation checks before the client actually tries to commit them, e.g. when they are put into the proposal store rather than when the client tries to commit them. I think this makes it easier for a client to deal with invalid proposals by other members.

Reply:
That's a good question. I tend to agree with you that doing the checks earlier would make more sense.

Reply:
@kkohbrok I agree, but just to make sure I understand your suggestion correctly:

1. we currently do checks in `CoreGroup::create_commit` (to prevent creating invalid proposals) and `CoreGroup::stage_commit` (to prevent accepting invalid proposals from other group members.)
2. you suggestion is to move the checks to `ProposalStore::add` so it works in both cases (we put our proposals and incoming proposals into the ProposalStore before committing them.)

Reply:
Sorry, I didn't make a clear suggestion of where exactly to put the check. Just to make sure we're on the same page, my understanding of our current way of doing validation is as follow:

1. It should be impossible for the consumer to not do the checks.
2. The checks should be at the level of the `CoreGroup` iff they are mandated by the spec and the `MlsGroup` otherwise.
3. `MlsGroup` should do as much of the work as possible for the user, where in the case of `CoreGroup`, we can expect the user to jump through a hoop or two.

We could do the check in `ProposalStore::add` as you suggest, but that would require passing all the information required for validation into the function. That would work, but it would bloat the otherwise relatively simple `ProposalStore` quite a bit. Another way would be to go the same route as for other `MlsMessages`, which would mean we create an intermediate type `UnverifiedProposal` (or `UnvalidatedProposal`), which allows a transformation to `QueuedProposal` that requires extra group-specific information and that does all the validation steps that we want. This would allow us to encapsulate the validation complexity somewhat and keep the `QueuedProposal`/`ProposalStore` types relatively simple. 

To force the user to perform this step, it would have to be impossible to circumvent the step from `MlsPlaintext` to `UnverifiedProposal`, which might pose problems in other places where we handle proposals.

Reply:
> Sorry, I didn't make a clear suggestion of where exactly to put the check.

Oops, I am sorry, I misread your earlier message. 😬

Re: argument against putting it in `ProposalStore::add`: that is a good point. Then it might be better to put the checks in [`StagedCommit::new`](https://github.com/openmls/openmls/blob/bd0c52f064dc8edc5063cbeeca224a592ac1d4fa/openmls/src/group/core_group/staged_commit.rs#L342-L352) since it would not overcomplicate the `ProposalStore` nor mess with existing flows.

Reply:
I talked to Raphael about where to put these checks and it turns out what I propose doesn't actually line up with what the spec requires 😅

Long story short, the checks are exactly right in the place that they are right now. Checking earlier is possible only in  a few cases and we might do so on the `MlsGroup` level for consumer convenience, but that's nothing we have to worry about in this PR.